### ISSUE TITLE
Compile sigfault error - Invalid memory access when compiling generic

### ISSUE THREAD

**Issue by Meldanor** [crystal-lang/crystal#6684]
Problem:

You get an invalid memory adress error when compiling a generic and wants to access Type#name

Error (from my machine):
```bash
Invalid memory access (signal 11) at address 0x0
[0x10d77cbb7] *CallStack::print_backtrace:Int32 +39
[0x10ce72bfe] __crystal_sigfault_handler +46
[0x7fff73e41f5a] _sigtramp +26
[0x10e57f5ef] _ZN4llvm11PointerType3getEPNS_4TypeEj +21
[0x10e1dcbff] _ZN4llvm17GetElementPtrInst16getGEPReturnTypeEPNS_4TypeEPNS_5ValueENS_8ArrayRefIS4_EE +63
[0x10e183eaf] _ZN4llvm17GetElementPtrInst6CreateEPNS_4TypeEPNS_5ValueENS_8ArrayRefIS4_EERKNS_5TwineEPNS_11InstructionE +111
[0x10dfad782] _ZN4llvm9IRBuilderINS_14ConstantFolderENS_24IRBuilderDefaultInserterEE17CreateInBoundsGEPEPNS_4TypeEPNS_5ValueENS_8ArrayRefIS7_EERKNS_5TwineE +114
[0x10e502592] LLVMBuildInBoundsGEP +59
[0x10d4e6eaa] *Crystal::CodeGenVisitor@Crystal::LLVMBuilderHelper#gep<LLVM::Value, Int32, Int32, String>:LLVM::Value +170
[0x10d519bd0] *Crystal::CodeGenVisitor#visit<Crystal::Assign>:(Bool | Nil) +1280
[0x10d500811] *Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::CodeGenVisitor>:Nil +1825
[0x10d5008a9] *Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::CodeGenVisitor>:Nil +1977
[0x10d514f3c] *Crystal::CodeGenVisitor#codegen_fun<String, Crystal::Def+, Crystal::Type+, Bool, Crystal::CodeGenVisitor::ModuleInfo, Bool, Bool>:LLVM::Function +4444
[0x10d517fdd] *Crystal::CodeGenVisitor#target_def_fun<Crystal::Def+, Crystal::Type+>:LLVM::Function +2445
[0x10d51f8ed] *Crystal::CodeGenVisitor#visit<Crystal::Call>:Bool +4525
[0x10d5029be] *Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::CodeGenVisitor>:Nil +10446
[0x10d5008a9] *Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::CodeGenVisitor>:Nil +1977
[0x10d514f3c] *Crystal::CodeGenVisitor#codegen_fun<String, Crystal::Def+, Crystal::Type+, Bool, Crystal::CodeGenVisitor::ModuleInfo, Bool, Bool>:LLVM::Function +4444
[0x10d517fdd] *Crystal::CodeGenVisitor#target_def_fun<Crystal::Def+, Crystal::Type+>:LLVM::Function +2445
[0x10d51f8ed] *Crystal::CodeGenVisitor#visit<Crystal::Call>:Bool +4525
[0x10d5029be] *Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::CodeGenVisitor>:Nil +10446
[0x10d524204] *Crystal::CodeGenVisitor#prepare_call_args_non_external<Crystal::Call, Crystal::Def+, Crystal::Type+>:Tuple(Array(LLVM::Value), Bool) +1140
[0x10d51f766] *Crystal::CodeGenVisitor#visit<Crystal::Call>:Bool +4134
[0x10d5029be] *Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::CodeGenVisitor>:Nil +10446
[0x10d502d52] *Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::CodeGenVisitor>:Nil +11362
[0x10d5008a9] *Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::CodeGenVisitor>:Nil +1977
[0x10d502cf0] *Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::CodeGenVisitor>:Nil +11264
[0x10d5008a9] *Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::CodeGenVisitor>:Nil +1977
[0x10d5051d1] *Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::CodeGenVisitor>:Nil +20705
[0x10d514f3c] *Crystal::CodeGenVisitor#codegen_fun<String, Crystal::Def+, Crystal::Type+, Bool, Crystal::CodeGenVisitor::ModuleInfo, Bool, Bool>:LLVM::Function +4444
[0x10d517fdd] *Crystal::CodeGenVisitor#target_def_fun<Crystal::Def+, Crystal::Type+>:LLVM::Function +2445
[0x10d51f8ed] *Crystal::CodeGenVisitor#visit<Crystal::Call>:Bool +4525
[0x10d5029be] *Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::CodeGenVisitor>:Nil +10446
[0x10d523df8] *Crystal::CodeGenVisitor#prepare_call_args_non_external<Crystal::Call, Crystal::Def+, Crystal::Type+>:Tuple(Array(LLVM::Value), Bool) +104
[0x10d51f766] *Crystal::CodeGenVisitor#visit<Crystal::Call>:Bool +4134
[0x10d5029be] *Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::CodeGenVisitor>:Nil +10446
[0x10d524204] *Crystal::CodeGenVisitor#prepare_call_args_non_external<Crystal::Call, Crystal::Def+, Crystal::Type+>:Tuple(Array(LLVM::Value), Bool) +1140
[0x10d51f766] *Crystal::CodeGenVisitor#visit<Crystal::Call>:Bool +4134
[0x10d5029be] *Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::CodeGenVisitor>:Nil +10446
[0x10d5008a9] *Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::CodeGenVisitor>:Nil +1977
[0x10d502d52] *Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::CodeGenVisitor>:Nil +11362
[0x10d5008a9] *Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::CodeGenVisitor>:Nil +1977
[0x10d514f3c] *Crystal::CodeGenVisitor#codegen_fun<String, Crystal::Def+, Crystal::Type+, Bool, Crystal::CodeGenVisitor::ModuleInfo, Bool, Bool>:LLVM::Function +4444
[0x10d5012a7] *Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::CodeGenVisitor>:Nil +4535
[0x10d519809] *Crystal::CodeGenVisitor#visit<Crystal::Assign>:(Bool | Nil) +313
[0x10d500811] *Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::CodeGenVisitor>:Nil +1825
[0x10d5008a9] *Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::CodeGenVisitor>:Nil +1977
[0x10d505b56] *Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::CodeGenVisitor>:Nil +23142
[0x10d520044] *Crystal::CodeGenVisitor#visit<Crystal::Call>:Bool +6404
[0x10d5029be] *Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::CodeGenVisitor>:Nil +10446
[0x10d5008a9] *Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::CodeGenVisitor>:Nil +1977
[0x10d505b56] *Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::CodeGenVisitor>:Nil +23142
[0x10d520044] *Crystal::CodeGenVisitor#visit<Crystal::Call>:Bool +6404
[0x10d5029be] *Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::CodeGenVisitor>:Nil +10446
[0x10d500c83] *Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::CodeGenVisitor>:Nil +2963
[0x10d5008a9] *Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::CodeGenVisitor>:Nil +1977
[0x10d4d234a] *Crystal::Compiler#codegen<Crystal::Program, Crystal::ASTNode+, Array(Crystal::Compiler::Source), String>:(Tuple(Array(Crystal::Compiler::CompilationUnit), Array(String)) | Nil) +490
[0x10d6f8bc4] *Crystal::Compiler#compile<Array(Crystal::Compiler::Source), String>:Crystal::Compiler::Result +116
[0x10ce85e9d] *Crystal::Command#run:(Bool | Crystal::Compiler::Result | Nil) +845
[0x10ce50bf4] main +212
```

Error (from crystal-play):
```bash
Invalid memory access (signal 11) at address 0x0
[0x56328dfca296] ???
[0x56328d6d4c5a] __crystal_sigfault_handler +42
[0x7fef38cfd3b0] ???
[0x7fef39e71354] _ZN4llvm11PointerType3getEPNS_4TypeEj +20
[0x7fef39dace1e] LLVMBuildInBoundsGEP +302
[0x56328dd2cfba] ???
[0x56328dd557c9] ???
[0x56328dd61622] ???
[0x56328dd47377] ???
[0x56328dd4744b] ???
[0x56328dd5c741] ???
[0x56328dd5fba6] ???
[0x56328dd6753e] ???
[0x56328dd49853] ???
[0x56328dd4744b] ???
[0x56328dd5c741] ???
[0x56328dd5fba6] ???
[0x56328dd6753e] ???
[0x56328dd49853] ???
[0x56328dd4744b] ???
[0x56328dd1763c] ???
[0x56328df41ad3] ???
[0x56328df9bf91] ???
[0x56328d6ee3ad] ???
[0x56328d6b69a7] __crystal_main +13591
[0x56328d6ba726] main +198
[0x7fef388ca003] __libc_start_main +243
[0x56328d6b33be] _start +46
[0x0] ???

```

Example code:
```crystal
class ErrorResponse
  def initialize(@statusCode : Int32)
  end

  def to_s(io)
    io << "Type: " << self.class.name << ";"
    io << "Status: " << self.@statusCode
  end
end

class EntityNotFound(T) < ErrorResponse
  def initialize
    @statusCode = 404
    @entityName = T.name
  end
end

EntityNotFound(String).new
```

Runable code: https://play.crystal-lang.org/#/r/4wwc

Crystal -v:

> Crystal 0.26.0 (2018-08-17)
>
>LLVM: 6.0.1
>Default target: x86_64-apple-macosx

uname -a:
> 17.7.0 Darwin Kernel Version 17.7.0: Thu Jun 21 22:53:14 PDT 2018; root:xnu-4570.71.2~1/RELEASE_X86_64 x86_64

**Comment by Meldanor** [comment]
I don't know if accessing the name of a type is allowed or possible (I know crystal only for 2 weeks), but the compile message is at least very hard to understand and to find the error.

**Comment by RX14** [comment]
This should compile.

**Comment by RX14** [comment]
Using LLVM compiled with debug assertions: 
```
crystal: ../include/llvm/IR/Instructions.h:830: llvm::Type* llvm::checkGEPType(llvm::Type*): Assertion `Ty && "Invalid GetElementPtrInst indices for type!"' failed
```

**Comment by RX14** [comment]
We seem to get:
```
instance_vars            # => {"@statusCode" => @statusCode, "@entityName" => @entityName}
superclass.instance_vars # => {"@statusCode" => @statusCode}
```
from this code which is causing this error.

**Comment by jhass** [comment]
Still reproduces with latest master:

```
Invalid memory access (signal 11) at address 0x0
[0x562b66a2e906] *CallStack::print_backtrace:Int32 +118
[0x562b66a15e49] __crystal_sigfault_handler +217
[0x562b67905a93] sigfault_handler +35
[0x7fd9fa2c34d0] ???
[0x7fd9faffc3f4] _ZN4llvm11PointerType3getEPNS_4TypeEj +20
[0x7fd9faf37ebe] LLVMBuildInBoundsGEP +302
[0x562b677b1b5a] *LLVM::Builder#inbounds_gep<LLVM::Value, LLVM::Value, LLVM::Value, String>:LLVM::Value +218
[0x562b677e0ae3] *Crystal::CrystalLLVMBuilder#inbounds_gep<LLVM::Value, LLVM::Value, LLVM::Value, String>:LLVM::Value +131
[0x562b6774150b] *Crystal::CodeGenVisitor +91
[0x562b6774149f] *Crystal::CodeGenVisitor +127
[0x562b67743468] *Crystal::CodeGenVisitor +72
[0x562b6774340f] *Crystal::CodeGenVisitor#aggregate_index<LLVM::Value, Int32>:LLVM::Value +47
[0x562b6775df2e] *Crystal::CodeGenVisitor#instance_var_ptr<Crystal::Type+, String, LLVM::Value>:LLVM::Value +1246
[0x562b6778be28] *Crystal::CodeGenVisitor#codegen_assign<Crystal::ASTNode+, Crystal::ASTNode+, Crystal::Assign>:(Bool | Nil) +840
[0x562b6778b5ca] *Crystal::CodeGenVisitor#visit<Crystal::Assign>:(Bool | Nil) +474
[0x562b66d0d18f] *Crystal::ASTNode+ +3631
[0x562b67758f41] *Crystal::CodeGenVisitor#accept<Crystal::ASTNode+>:Nil +17
[0x562b6778aa24] *Crystal::CodeGenVisitor#visit<Crystal::Expressions>:Bool +276
[0x562b66d0ca19] *Crystal::ASTNode+ +1721
[0x562b67758f41] *Crystal::CodeGenVisitor#accept<Crystal::ASTNode+>:Nil +17
[0x562b6776588e] *Crystal::CodeGenVisitor#codegen_fun<String, Crystal::Def+, Crystal::Type+, Bool, Crystal::CodeGenVisitor::ModuleInfo, Bool, Bool>:LLVM::Function +3566
[0x562b6777de88] *Crystal::CodeGenVisitor#codegen_fun<String, Crystal::Def+, Crystal::Type+>:LLVM::Function +184
[0x562b6777d771] *Crystal::CodeGenVisitor#target_def_fun<Crystal::Def+, Crystal::Type+>:LLVM::Function +417
[0x562b67780c56] *Crystal::CodeGenVisitor#codegen_call<Crystal::Call, Crystal::Def+, Crystal::Type+, Array(LLVM::Value)>:(Bool | LLVM::Value | Nil) +326
[0x562b677764ce] *Crystal::CodeGenVisitor#visit<Crystal::Call>:Bool +670
[0x562b66d0d7a1] *Crystal::ASTNode+ +5185
[0x562b67758f41] *Crystal::CodeGenVisitor#accept<Crystal::ASTNode+>:Nil +17
[0x562b6778aa24] *Crystal::CodeGenVisitor#visit<Crystal::Expressions>:Bool +276
[0x562b66d0ca19] *Crystal::ASTNode+ +1721
[0x562b67758f41] *Crystal::CodeGenVisitor#accept<Crystal::ASTNode+>:Nil +17
[0x562b6776588e] *Crystal::CodeGenVisitor#codegen_fun<String, Crystal::Def+, Crystal::Type+, Bool, Crystal::CodeGenVisitor::ModuleInfo, Bool, Bool>:LLVM::Function +3566
[0x562b6777de88] *Crystal::CodeGenVisitor#codegen_fun<String, Crystal::Def+, Crystal::Type+>:LLVM::Function +184
[0x562b6777d771] *Crystal::CodeGenVisitor#target_def_fun<Crystal::Def+, Crystal::Type+>:LLVM::Function +417
[0x562b67780c56] *Crystal::CodeGenVisitor#codegen_call<Crystal::Call, Crystal::Def+, Crystal::Type+, Array(LLVM::Value)>:(Bool | LLVM::Value | Nil) +326
[0x562b677764ce] *Crystal::CodeGenVisitor#visit<Crystal::Call>:Bool +670
[0x562b66d0d7a1] *Crystal::ASTNode+ +5185
[0x562b67758f41] *Crystal::CodeGenVisitor#accept<Crystal::ASTNode+>:Nil +17
[0x562b6778aa24] *Crystal::CodeGenVisitor#visit<Crystal::Expressions>:Bool +276
[0x562b66d0ca19] *Crystal::ASTNode+ +1721
[0x562b67758f41] *Crystal::CodeGenVisitor#accept<Crystal::ASTNode+>:Nil +17
[0x562b66b7b3af] *Crystal::Program#codegen<Crystal::ASTNode+, (Bool | Crystal::Compiler::EmitTarget | Nil), Crystal::Debug>:Hash(String, Crystal::CodeGenVisitor::ModuleInfo)+ +95
[0x562b66b7b34a] *Crystal::Program#codegen:debug:single_module<Crystal::ASTNode+, Crystal::Debug, (Bool | Crystal::Compiler::EmitTarget | Nil)>:Hash(String, Crystal::CodeGenVisitor::ModuleInfo)+ +74
[0x562b676c2607] *Crystal::Compiler#codegen<Crystal::Program, Crystal::ASTNode+, Array(Crystal::Compiler::Source), String>:(Tuple(Array(Crystal::Compiler::CompilationUnit), Array(String)) | Nil) +487
[0x562b676c5be9] *Crystal::Compiler#compile<Array(Crystal::Compiler::Source), String>:Crystal::Compiler::Result +185
[0x562b67856b39] *Crystal::Command::CompilerConfig#compile<String>:Crystal::Compiler::Result +57
[0x562b67856ae4] *Crystal::Command::CompilerConfig#compile:Crystal::Compiler::Result +36
```

**Comment by straight-shoota** [comment]
The error no longer reproduces. It appears this has been fixed pre 1.0.
